### PR TITLE
Feature Extractor Refactor

### PR DESCRIPTION
**Pull request by ashishbdatta**
# Description

- This PR changes the default behavior of `FeatureExtractor` class to leverage models from `timm` rather than `torchvision`. By using `timm` we can easily access the features we care about rather than relying on `forward_hooks`

## Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

## Checklist

- [X] My code follows the [pre-commit style and check guidelines](https://openvinotoolkit.github.io/anomalib/guides/using_pre_commit.html#pre-commit-hooks) of this project.
- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes

### CODE REVIEW

**Comment by ashishbdatta** [comment]
I made an assumption [here](https://github.com/ashishbdatta/anomalib/blob/ad/feature-extractor/anomalib/models/components/feature_extractors/feature_extractor.py#L52) that the layers will follow the format of `layer<num_of_layer` like [here]( https://github.com/openvinotoolkit/anomalib/blob/a9461407e0f27de440daef5d7d5e1f433fec1bda/anomalib/models/padim/config.yaml#L28-L30) 

But I can see the `DFKDE` model uses the key `avgpool`. What is the preference here? to support anyname or restrict the config to `layer<num_of_layer>` convention?

**Comment by ashishbdatta** [comment]
Please let me know if this looks good @samet-akcay

**Comment by samet-akcay** [comment]
> Please let me know if this looks good @samet-akcay  

@ashishbdatta, there will probably be merge conflicts due to #453. I'll review it anyway, and let you know

**Comment by ashwinvaidya17** [comment]
> I made an assumption [here](https://github.com/ashishbdatta/anomalib/blob/ad/feature-extractor/anomalib/models/components/feature_extractors/feature_extractor.py#L52) that the layers will follow the format of `layer<num_of_layer>` like [here](https://github.com/openvinotoolkit/anomalib/blob/a9461407e0f27de440daef5d7d5e1f433fec1bda/anomalib/models/padim/config.yaml#L28-L30) since so many configs use this convention
> 
> But I can see the `DFKDE` model uses the key `avgpool`. What is the preference here? to support anyname or restrict the config to `layer<num_of_layer>` convention?

I'd be in the favour of supporting any layer but it might be tricky with this new design as timm requires layer indices. Because the model isn't initialized, it is hard to know the layer id which corresponds to the name. Here is an idea - Instead of splitting based on `layer` maybe we can update the config to reflect

```yaml
 layers:
    - block 1: 1
    - block 2: 2
    - block 3: 3
```

```yaml
 layers:
    - avgpool: 4
```

Since we only need the id, the layer key here might be unnecessary, it would make it readable for the users. As this will be a breaking change we can club this with the CLI upgrade and make it part of the 0.4.x release.

**Comment by ashishbdatta** [comment]
We could do something like

```python
def _map_layer_to_idx(self) -> List[int]:
      idx = []
      # Create temp model to search child nodes
      features = timm.create_model(
                        backbone,
                        pretrained,
                        features_only,
                        )
      for i in self.layers
           # Search children nodes and return index they are found at
           idx.append(list(dict(features.named_children()).keys()).index(i))
      return idx
```

Only problem is that this would return indices of `[4, 5, 6]` whereas we need indices of `[1, 2, 3]` because `timm` for some reason ignores non-`nn.Sequential` blocks for selecting indices. 

For e.g., in a `Resnet18` model they ignore `conv1, bn1, maxpool1` in their indexing.

File: configs/model/dfkde.yaml [hunk 1]
```diff
@@ -25,7 +25,7 @@ model:
     backbone: resnet18
     pre_trained: true
     layers:
-      - avgpool
+      - layer3
```
Comment (ashwinvaidya17) [inline configs/model/dfkde.yaml:28]: layer4

**Review by ashwinvaidya17 — CHANGES_REQUESTED** [review]
The tests should pass with these changes.

**Review by ashwinvaidya17 — APPROVED** [review]
Thanks for migrating the feature extractor.

**Comment by samet-akcay** [comment]
Thanks @ashishbdatta. Looking good!

**Review by samet-akcay — APPROVED** [review]

**Pull request merged**